exercises/practice/robot-name/.approaches/mass-name-generation/content.md

```diff
@@ -0,0 +1,42 @@
+# Mass Name Generation
+We'd first have to generate all the possible names, shuffle them, and then use `next` (the simplest way) or maintain a `current_index` and get the name.
+Note that selecting randomly from the list of all names would be incorrect, as there's a possibility of the name being repeated.
+
+Here's a possible way to do it:
+
+```python
+from itertools import product
+from random import shuffle
+from string import ascii_uppercase
+
+letter_pairs = (''.join(p) for p in product(ascii_uppercase, ascii_uppercase))
+numbers = (str(i).zfill(3) for i in range(1000))
+names = [l + n for l, n in product(letter_pairs, numbers)]
+
+shuffle(names)
+NAMES = iter(names)
+
+class Robot(object):
+    def __init__(self):
+        self.reset()
+    def reset(self):
+        self.name = next(NAMES)
+```
+
+The first few lines of the mass name generation uses [`itertools.product`][itertools-product].
+The resultant code is a simplification of:
+```python
+letter_pairs = (''.join((l1, l2)) for l1 in ascii_uppercase for l2 in ascii_uppercase)
+numbers = (str(i).zfill(3) for i in range(1000))
+names = [l + n for l in letter_pairs for n in numbers]
+```
+
+After the name generation, the names are shuffled - using the [default `seed`][random-seed] in the `random` module (the current timestamp). 
+When the tests reseed `random`, this has no effect as the names were shuffled before that.
+
+We then set `NAMES` to the iterable of names, and in `reset`, set the robot's name to the  `next(name)`. 
+If you'd like, read more on [`iter` and `next`][iter-and-next].
```

Comment:
It might be worth discussing the pros and cons/drawbacks of the different approaches. The mass generation has a high start up cost and memory cost. It's hugely inefficient if we're going to generate a relatively small number of names.

Conversely, the "on the fly" approach can be very expensive for a large number of names, since collisions occur more often as the number of available names gets smaller.

Reply:
Shouldn't that come under articles instead of approaches? Perhaps a follow-on PR?

Reply:
> I like the idea, but shouldn't that come under articles instead of approaches?

The approaches aren't just a list of possible solutions; they are meant to dig deeper into the problem and different approaches to solving the problem. I think it makes sense to discuss the pros and cons of approaches alongside the approach. When describing "you can solve it like this", it is good to provide context and details, like why or when you would want to solve it this way. 

> Perhaps a follow-on PR?

Whatever works for you.

Reply:
I've committed, but out of curiosity, what's the difference between approaches and articles? I thought articles were supposed to compare approaches' performance or other interesting stuff like that, while approaches were meant to detail how an exercise could be solved in a certain manner.

Reply:
@safwansamsudeen -- you are correct.  Here are the [approaches docs](https://exercism.org/docs/building/tracks/approaches).  Here are the [articles docs](https://exercism.org/docs/building/tracks/articles).

But also:  I have not yet had time to review this.  I thought you were going to put this on hold?

Reply:
Article is definitely better suited for in depth comparison/exploration. For approaches, the docs say,

> Feel free to dig deep into the topic

I wouldn't expect an extensive comparison of approaches in the approach docs, but I think a line or three giving a high level overview of the pros and cons of the approach is appropriate.